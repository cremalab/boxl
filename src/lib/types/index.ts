import React, { CSSProperties } from "react";
import {
  AnyStyledComponent,
  CSSObject,
  InterpolationFunction,
  ThemeProps,
} from "styled-components";

export type BoxlComponentType<P, T, E> = React.Component<BoxlProps<P, T, E>>;

export type BoxlThunkReturn<E extends BoxlPropElement, T = {}> = <
  P,
  D extends BoxlPropsPartial<P, T, EE> = BoxlPropsPartial<P, T, EE>,
  EE = JSX.IntrinsicElements[E]
>(
  d?: D | undefined
) => {
  (p: BoxlProps<P, T, EE>): React.ReactElement<BoxlProps<P, T, EE>>;
  defaultProps: D | undefined;
};

export type BoxlTwo<T = {}> = { [E in BoxlPropElement]: BoxlThunkReturn<E, T> };

export type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;
export type Entry<O, P extends keyof O = keyof O> = [P, O[P]];

export type BoxlElement<E> = E;

export type BoxlProp<A, P, T> =
  | (A | undefined)
  | (BoxlPropThemeFn<A | undefined, P, T>)
  | (BoxlPropMediaQuery<A | undefined, P, T>);

export type BoxlPropAlignmentHorizontal<P, T> = BoxlProp<
  "left" | "center" | "right",
  P,
  T
>;

export type BoxlPropAlignmentVertical<P, T> = BoxlProp<
  "top" | "center" | "bottom",
  P,
  T
>;

export type BoxlPropDirection<P, T> = BoxlProp<"horizontal" | "vertical", P, T>;

export type BoxlPropElement = keyof JSX.IntrinsicElements;

export type BoxlPropComponent<P> = React.ComponentType<P>;

export type BoxlPropGrow<P, T> = BoxlProp<number, P, T>;

export type BoxlPropIdealSize<P, T> = BoxlProp<string, P, T>;

export type BoxlPropMediaQuery<A, P, T> = {
  [key: string]: A | BoxlPropThemeFn<A, P, T>;
};

export type BoxlPropPadding<P, T> = BoxlProp<string | number, P, T>;

export type BoxlPropSpacing<P, T> = BoxlProp<string | number, P, T>;

export type BoxlPropStyle<P, T> =
  | string
  | CSSObject
  | ReadonlyArray<string> & CSSProperties
  | ((styleFn: BoxlThemeThunk<P, T>) => BoxlThemeThunkReturn<P, T>);

export type BoxlPropThemeFn<A, P, T> = (
  props: BoxlPropsBaseThemed<P, T>
) => A | BoxlPropMediaQuery<A, P, T>;

export type BoxlPropWrap<P, T> = BoxlProp<"auto" | "even", P, T>;

export type BoxlThemeThunk<P = {}, T = {}> = (
  strings: ReadonlyArray<string>,
  ...interpolations: Array<
    InterpolationFunction<BoxlPropsBaseThemed<P, T>> | string
  >
) => BoxlThemeThunkReturn<P, T>;

export type BoxlThemeThunkReturn<P, T> = {
  literals: ReadonlyArray<string>;
  interpolations: Array<
    InterpolationFunction<BoxlPropsBaseThemed<P, T>> | string
  >;
};

export interface BoxlPropsBase<P = {}, T = {}> {
  alignHorizontal?: BoxlPropAlignmentHorizontal<P, T>;
  alignVertical?: BoxlPropAlignmentVertical<P, T>;
  childGrow?: BoxlPropGrow<P, T>;
  childIdealSize?: BoxlPropIdealSize<P, T>;
  childWrap?: BoxlPropWrap<P, T>;
  children?: React.ReactNode;
  component?: BoxlPropComponent<P>;
  direction?: BoxlPropDirection<P, T>;
  element?: BoxlPropElement;
  grow?: BoxlPropGrow<P, T>;
  idealSize?: BoxlPropIdealSize<P, T>;
  isChild?: boolean;
  isDummy?: boolean;
  padding?: BoxlPropPadding<P, T>;
  propsParent?: BoxlPropsBase<P, T>;
  spacing?: BoxlPropSpacing<P, T>;
  style?: BoxlPropStyle<P, T>;
}

export type BoxlPropsBaseThemed<P, T> = BoxlPropsBase<P, T> & ThemeProps<T> & P;

export type BoxlStyledComponent = AnyStyledComponent;

export type BoxlComponentProps<P = {}, T = {}> = {
  boxlProps?: BoxlPropsBase<P, T>;
} & P;

export type BoxlProps<P = {}, T = {}, E = {}> = BoxlPropsBase<P, T> &
  BoxlElement<E> &
  P;

export type BoxlPropsThemed<T, P = {}, E = {}> = BoxlProps<P, T, E>;

export type BoxlPropsPartial<P = {}, T = {}, E = {}> = Partial<
  BoxlProps<P, T, E>
>;

export type InferPropsType<T> = T extends (...args: infer Args) => infer _R
  ? Args[0]
  : T extends React.ComponentClass<infer P>
    ? P
    : T extends keyof JSX.IntrinsicElements ? JSX.IntrinsicElements[T] : never;
