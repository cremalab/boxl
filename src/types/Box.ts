import * as React from "react";
import { InterpolationFunction, ThemeProps } from "styled-components";
export { Box } from "../web";

type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;

export type BoxPropThemeFn<A, T> = (
  props: BoxThemeProps<T>
) => A | BoxPropMediaQuery<A, T>;

export interface BoxPropMediaQuery<A, T> {
  [key: string]: A | BoxPropThemeFn<A, T>;
}

export type BoxProp<A, T> = A | BoxPropMediaQuery<A, T> | BoxPropThemeFn<A, T>;

export type BoxSpacingInfo = string;

export type BoxPropDirection<T> = BoxProp<"horizontal" | "vertical", T>;
export type BoxPropGrow<T> = BoxProp<number, T>;
export type BoxPropPadding<T> = BoxProp<string | number | undefined, T>;
export type BoxPropSpacing<T> = BoxProp<string | number | undefined, T>;
export type BoxPropWrap<T> = BoxProp<"auto" | "even" | undefined, T>;
export type BoxPropIdealWidth<T> = BoxProp<string, T>;
export type BoxPropHorizontalAlignment<T> = BoxProp<
  "left" | "center" | "right" | undefined,
  T
>;
export type BoxPropVerticalAlignment<T> = BoxProp<
  "top" | "center" | "bottom" | undefined,
  T
>;

export type BoxThemeThunk<T> = (
  strings: ReadonlyArray<string>,
  ...interpolations: Array<InterpolationFunction<BoxThemeProps<T>> | string>
) => BoxThemeThunkReturn<T>;

export interface BoxThemeThunkReturn<T> {
  literals: ReadonlyArray<string>;
  interpolations: Array<InterpolationFunction<BoxThemeProps<T>> | string>;
}

export type BoxPropStyle<T> =
  | ((styleFn: BoxThemeThunk<T>) => BoxThemeThunkReturn<T>)
  | (string);

export interface BoxContainerProps<T> {
  grow: BoxPropGrow<T>;
  padding?: BoxPropPadding<T>;
  styleString?: BoxPropStyle<T>;
}

export interface BoxChildrenProps<T> {
  childWrap?: BoxPropWrap<T>;
  grow?: BoxPropGrow<T>;
  horizontalAlign?: BoxPropHorizontalAlignment<T>;
  iDirection: BoxPropDirection<T>;
  padding?: BoxPropPadding<T>;
  spacingInfo?: BoxPropSpacing<T>;
  styleString?: BoxPropStyle<T>;
  verticalAlign?: BoxPropVerticalAlignment<T>;
}

export interface BoxChildProps<T> {
  grow: BoxPropGrow<T>;
  idealWidth?: BoxPropIdealWidth<T>;
  isDummy: boolean;
  spacingInfo?: BoxPropSpacing<T>;
}

export type BoxElement = keyof React.ReactHTML;
export type El = Omit<React.AllHTMLAttributes<HTMLElement>, "style">;

export interface BoxProps<T> extends El {
  childGrow?: BoxPropGrow<T>;
  childIdealWidth?: BoxPropIdealWidth<T>;
  childWrap?: BoxPropWrap<T>;
  direction?: BoxPropDirection<T>;
  element?: BoxElement;
  grow?: BoxPropGrow<T>;
  horizontalAlign?: BoxPropHorizontalAlignment<T>;
  idealWidth?: BoxPropIdealWidth<T>;
  isChild?: boolean;
  padding?: BoxPropPadding<T>;
  spacing?: BoxPropSpacing<T>;
  style?: BoxPropStyle<T>;
  verticalAlign?: BoxPropVerticalAlignment<T>;
}

export type BoxThemeProps<T> = BoxProps<T> & ThemeProps<T>;
