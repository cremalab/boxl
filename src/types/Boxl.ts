import * as React from "react";
import { InterpolationFunction, ThemeProps } from "styled-components";
export * from "..";

export type BoxlChildProps<P, T> = {
  grow: BoxlPropGrow<P, T>;
  idealWidth?: BoxlPropIdealWidth<P, T>;
  isDummy: boolean;
};

export interface BoxlNativeProps<P = {}, T = {}> {
  alignHorizontal?: BoxlPropAlignmentHorizontal<P, T>;
  alignVertical?: BoxlPropAlignmentVertical<P, T>;
  childGrow?: BoxlPropGrow<P, T>;
  childIdealWidth?: BoxlPropIdealWidth<P, T>;
  childWrap?: BoxlPropWrap<P, T>;
  direction?: BoxlPropDirection<P, T>;
  element?: BoxlPropElement;
  grow?: BoxlPropGrow<P, T>;
  idealWidth?: BoxlPropIdealWidth<P, T>;
  isChild?: boolean;
  padding?: BoxlPropPadding<P, T>;
  spacing?: BoxlPropSpacing<P, T>;
  style?: BoxlPropStyle<P, T>;
  forwardedRef?: any;
}

export type BoxlComponentProps<P = {}, T = {}> = BoxlNativeProps<P, T> &
  BoxlElement &
  P;

export type BoxlElement = Omit<React.AllHTMLAttributes<HTMLElement>, "style">;

export type BoxlProp<A, P, T> =
  | A
  | (BoxlPropThemeFn<A, P, T>)
  | (BoxlPropMediaQuery<A, P, T>);

export type BoxlPropAlignmentHorizontal<P, T> = BoxlProp<
  "left" | "center" | "right" | undefined,
  P,
  T
>;

export type BoxlPropAlignmentVertical<P, T> = BoxlProp<
  "top" | "center" | "bottom" | undefined,
  P,
  T
>;

export type BoxlPropDirection<P, T> = BoxlProp<"horizontal" | "vertical", P, T>;

export type BoxlPropElement = keyof React.ReactHTML;

export type BoxlPropGrow<P, T> = BoxlProp<number | undefined, P, T>;

export type BoxlPropIdealWidth<P, T> = BoxlProp<string, P, T>;

export type BoxlPropMediaQuery<A, P, T> = {
  [key: string]: A | BoxlPropThemeFn<A, P, T>;
};

export type BoxlPropPadding<P, T> = BoxlProp<string | number | undefined, P, T>;

export type BoxlPropSpacing<P, T> = BoxlProp<string | number | undefined, P, T>;

export type BoxlPropStyle<P, T> =
  | string
  | ReadonlyArray<string>
  | ((styleFn: BoxlThemeThunk<P, T>) => BoxlThemeThunkReturn<P, T>);

export type BoxlPropThemeFn<A, P, T> = (
  props: BoxlPropsThemed<P, T>
) => A | BoxlPropMediaQuery<A, P, T>;

export type BoxlPropWrap<P, T> = BoxlProp<"auto" | "even" | undefined, P, T>;

export type BoxlProps<P = {}, T = {}> = BoxlComponentProps<P, T>;
export type BoxlPropsPartial<P, T> = BoxlProps<Partial<P>, T>;
export type BoxlPropsThemed<P, T> = BoxlProps<P, T> & ThemeProps<T>;

export type BoxlThemeThunk<P, T> = (
  strings: ReadonlyArray<string>,
  ...interpolations: Array<
    InterpolationFunction<BoxlPropsThemed<P, T>> | string
  >
) => BoxlThemeThunkReturn<P, T>;

export type BoxlThemeThunkReturn<P, T> = {
  literals: ReadonlyArray<string>;
  interpolations: Array<InterpolationFunction<BoxlPropsThemed<P, T>> | string>;
};

export type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;
