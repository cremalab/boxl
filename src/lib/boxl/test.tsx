import "jest-styled-components";
import React, { SFC } from "react";
import TestRenderer from "react-test-renderer";
import { cleanup, fireEvent, render } from "react-testing-library";
import { boxl as b } from ".";
import { BoxlComponentProps } from "../types";

describe("boxl", () => {
  const Boxl = b.div();

  describe("structure", () => {
    it("renders correct structure when no spacing", () => {
      const received = TestRenderer.create(<Boxl>Text</Boxl>);
      expect(received).toMatchSnapshot();
    });

    it("renders correct structure when spacing BUT children length === 1", () => {
      const received = TestRenderer.create(<Boxl spacing="1em">Text</Boxl>);
      expect(received).toMatchSnapshot();
    });

    it("renders correct structure when spacing AND children length > 1", () => {
      const received = TestRenderer.create(
        <Boxl spacing="1em">
          <div>Text 1</div>
          <div>Text 2</div>
        </Boxl>
      );
      expect(received).toMatchSnapshot();
    });

    it("renders correct structure when spacing AND children length > 1 and children are Boxl", () => {
      const received = TestRenderer.create(
        <Boxl spacing="1em">
          <Boxl>Text 1</Boxl>
          <Boxl>Text 2</Boxl>
        </Boxl>
      );
      expect(received).toMatchSnapshot();
    });

    it("renders defined `element`", () => {
      const received = TestRenderer.create(<Boxl element="a" />);
      expect(received).toMatchSnapshot();
    });

    it("renders defined `component`", () => {
      const P: SFC<BoxlComponentProps> = ({ boxlProps, ...props }) => (
        <p {...props} />
      );
      const B = b(P)();
      const received = TestRenderer.create(<B />);
      expect(received).toMatchSnapshot();
    });
  });

  describe("edge-case", () => {
    it("handles null child in children array", () => {
      const received = TestRenderer.create(
        <Boxl grow={1} spacing={"10px"} children={[null]} />
      );
      expect(received).toMatchSnapshot();
    });

    describe("naked text node", () => {
      it("is not wrapped when spacing undefined", () => {
        const received = TestRenderer.create(
          <Boxl alignVertical="center" grow={1}>
            <div key="1">1</div>2<div key="3">3</div>
          </Boxl>
        );
        expect(received).toMatchSnapshot();
      });

      it("is wrapped when spacing defined", () => {
        const received = TestRenderer.create(
          <Boxl alignVertical="center" grow={1} spacing={"10px"}>
            <div key="1">1</div>2<div key="3">3</div>
          </Boxl>
        );
        expect(received).toMatchSnapshot();
      });
    });
  });

  describe("direction", () => {
    it("default", () => {
      const received = TestRenderer.create(
        <Boxl>
          <div>1</div>
          <div>2</div>
        </Boxl>
      );
      expect(received).toMatchSnapshot();
    });

    it("vertical", () => {
      const received = TestRenderer.create(
        <Boxl direction="horizontal">
          <div>1</div>
          <div>2</div>
        </Boxl>
      );
      expect(received).toMatchSnapshot();
    });

    it("horizontal", () => {
      const received = TestRenderer.create(
        <Boxl direction="horizontal">
          <div>1</div>
          <div>2</div>
        </Boxl>
      );
      expect(received).toMatchSnapshot();
    });
  });

  describe("childWrap", () => {
    it("auto", () => {
      const received = TestRenderer.create(
        <Boxl direction="horizontal" childWrap="auto" childIdealSize="200px">
          <div>1</div>
          <div>2</div>
        </Boxl>
      );
      expect(received).toMatchSnapshot();
    });

    it("even", () => {
      const received = TestRenderer.create(
        <Boxl direction="horizontal" childWrap="even" childIdealSize="200px">
          <div>1</div>
          <div>2</div>
        </Boxl>
      );
      expect(received).toMatchSnapshot();
    });
  });

  describe("rerender", () => {
    afterEach(cleanup);

    it("maintains element instance between renders", () => {
      const Container = b.form({ spacing: "1em" });
      const Input = b.input();
      class Form extends React.Component {
        public state = {
          value: "",
        };
        public render() {
          return (
            <Container>
              <Input
                placeholder="test1"
                value={this.state.value}
                onChange={(e: any) =>
                  this.setState({ value: (e.target as any).value })
                }
              />
              <Input
                placeholder="test2"
                value={this.state.value}
                onChange={(e: any) =>
                  this.setState({ value: (e.target as any).value })
                }
              />
            </Container>
          );
        }
      }
      const { getByPlaceholderText } = render(<Form />);
      const input1 = getByPlaceholderText("test1") as HTMLInputElement;
      const value1 = "Great advice, I love your posts!";
      fireEvent.change(input1, {
        target: { value: value1 },
      });
      const input2 = getByPlaceholderText("test2") as HTMLInputElement;
      const value2 = "My password is PASSWORD";
      fireEvent.change(input2, {
        target: { value: value2 },
      });

      expect(input1).toMatchSnapshot();
      expect(input2).toMatchSnapshot();
    });
  });

  describe("prop:", () => {
    describe("alignHorizontal", () => {
      it("default", () => {
        const received = TestRenderer.create(<Boxl />);
        expect(received).toMatchSnapshot();
      });

      it("left", () => {
        const received = TestRenderer.create(<Boxl alignHorizontal="left" />);
        expect(received).toMatchSnapshot();
      });

      it("center", () => {
        const received = TestRenderer.create(<Boxl alignHorizontal="center" />);
        expect(received).toMatchSnapshot();
      });

      it("right", () => {
        const received = TestRenderer.create(<Boxl alignHorizontal="right" />);
        expect(received).toMatchSnapshot();
      });
    });

    describe("alignVertical", () => {
      it("default", () => {
        const received = TestRenderer.create(<Boxl />);
        expect(received).toMatchSnapshot();
      });

      it("top", () => {
        const received = TestRenderer.create(<Boxl alignVertical="top" />);
        expect(received).toMatchSnapshot();
      });

      it("center", () => {
        const received = TestRenderer.create(<Boxl alignVertical="center" />);
        expect(received).toMatchSnapshot();
      });

      it("bottom", () => {
        const received = TestRenderer.create(<Boxl alignVertical="bottom" />);
        expect(received).toMatchSnapshot();
      });
    });

    describe("childGrow", () => {
      it("works", () => {
        const received = TestRenderer.create(
          <Boxl childGrow={1}>
            <div>1</div>
            <div>2</div>
          </Boxl>
        );
        expect(received).toMatchSnapshot();
      });
    });

    describe("childIdealSize", () => {
      it("with direction vertical (no flex-basis)", () => {
        const received = TestRenderer.create(
          <Boxl childIdealSize="50%">
            <div>1</div>
            <div>2</div>
          </Boxl>
        );
        expect(received).toMatchSnapshot();
      });
      it("with direction horizontal (adds flex-basis)", () => {
        const received = TestRenderer.create(
          <Boxl direction="horizontal" childIdealSize="50%">
            <div>1</div>
            <div>2</div>
          </Boxl>
        );
        expect(received).toMatchSnapshot();
      });
    });

    describe("childWrap", () => {
      it("vertical", () => {
        const received = TestRenderer.create(
          <Boxl childWrap="auto">
            <Boxl>1</Boxl>
            <Boxl>2</Boxl>
          </Boxl>
        );
        expect(received).toMatchSnapshot();
      });

      it("auto", () => {
        const received = TestRenderer.create(
          <Boxl direction="horizontal" childWrap="auto">
            <div>1</div>
            <div>2</div>
          </Boxl>
        );
        expect(received).toMatchSnapshot();
      });

      it("even", () => {
        const received = TestRenderer.create(
          <Boxl direction="horizontal" childWrap="even">
            <div>1</div>
            <div>2</div>
          </Boxl>
        );
        expect(received).toMatchSnapshot();
      });

      it("even:spacing", () => {
        const received = TestRenderer.create(
          <Boxl direction="horizontal" childWrap="even" spacing="2px">
            <div>1</div>
            <div>2</div>
          </Boxl>
        );
        expect(received).toMatchSnapshot();
      });

      it("even:spacing:childIdealSize", () => {
        const received = TestRenderer.create(
          <Boxl
            direction="horizontal"
            childWrap="even"
            spacing="2px"
            childIdealSize="100px"
          >
            <div>1</div>
            <div>2</div>
          </Boxl>
        );
        expect(received).toMatchSnapshot();
      });
    });

    describe("component", () => {
      it("no argument w/ spacing", () => {
        const B = b()({ spacing: "1em" });
        const received = TestRenderer.create(
          <B direction="horizontal">
            <div>1</div>
            <div>2</div>
          </B>
        );
        expect(received).toMatchSnapshot();
      });

      it("no argument no spacing", () => {
        const B = b()();
        const received = TestRenderer.create(
          <B direction="horizontal">
            <div>1</div>
            <div>2</div>
          </B>
        );
        expect(received).toMatchSnapshot();
      });

      it("function", () => {
        const P: SFC<BoxlComponentProps<{ color: string }>> = ({
          boxlProps,
          ...props
        }) => <p {...props} />;
        const B = b(P)();
        const received = TestRenderer.create(
          <B color="red" direction="horizontal" />
        );
        expect(received).toMatchSnapshot();
      });

      it("function w/ spacing", () => {
        const P: SFC<BoxlComponentProps<{ color: string }>> = ({
          boxlProps,
          ...props
        }) => <p {...props} />;
        const B = b(P)({ spacing: "1em" });
        const received = TestRenderer.create(
          <B color="red" direction="horizontal">
            <div>1</div>
            <div>2</div>
          </B>
        );
        expect(received).toMatchSnapshot();
      });

      it("class", () => {
        // tslint:disable-next-line
        class P extends React.Component<BoxlComponentProps<{ color: string }>> {
          public render() {
            const { boxlProps, ...props } = this.props;
            return <p {...props} />;
          }
        }
        const B = b(P)<{ test: boolean }>();
        const received = TestRenderer.create(
          <B color="red" test={false} direction="horizontal" />
        );
        expect(received).toMatchSnapshot();
      });
    });

    describe("direction", () => {
      it("horizontal", () => {
        const received = TestRenderer.create(
          <Boxl direction="horizontal">
            <div>1</div>
          </Boxl>
        );
        expect(received).toMatchSnapshot();
      });

      it("vertical", () => {
        const received = TestRenderer.create(
          <Boxl direction="vertical">
            <div>1</div>
          </Boxl>
        );
        expect(received).toMatchSnapshot();
      });
    });

    describe("element", () => {
      it("a", () => {
        const A = b.a();
        const received = TestRenderer.create(
          <A href="google.com">google.com</A>
        );
        expect(received).toMatchSnapshot();
      });
    });

    describe("grow", () => {
      it("child boxl has flex-grow: 1 by default when full structure", () => {
        const received = TestRenderer.create(
          <Boxl spacing="2px">
            <Boxl />
            <Boxl spacing="4px">
              <Boxl />
              <Boxl />
            </Boxl>
          </Boxl>
        );
        expect(received).toMatchSnapshot();
      });

      it("child grow is set on BoxlChild", () => {
        const received = TestRenderer.create(
          <Boxl spacing="2px">
            <Boxl />
            <Boxl grow={1} />
          </Boxl>
        );
        expect(received).toMatchSnapshot();
      });
    });

    describe("idealSize", () => {
      it("works", () => {
        const received = TestRenderer.create(<Boxl idealSize="100px" />);
        expect(received).toMatchSnapshot();
      });

      it("child idealSize is set on BoxlChild", () => {
        const received = TestRenderer.create(
          <Boxl spacing="2px">
            <Boxl idealSize="100px" />
            <Boxl grow={1} />
          </Boxl>
        );
        expect(received).toMatchSnapshot();
      });
    });

    describe("spacing", () => {
      it("no children", () => {
        const received = TestRenderer.create(<Boxl spacing="1em" />);
        expect(received).toMatchSnapshot();
      });

      it("one child", () => {
        const received = TestRenderer.create(
          <Boxl spacing="1em">
            <div />
          </Boxl>
        );
        expect(received).toMatchSnapshot();
      });

      it("> 1 child", () => {
        const received = TestRenderer.create(
          <Boxl spacing="1em">
            <div />
            <div />
          </Boxl>
        );
        expect(received).toMatchSnapshot();
      });
    });

    describe("css", () => {
      it("string", () => {
        const received = TestRenderer.create(<Boxl css="color: red;" />);
        expect(received).toMatchSnapshot();
      });

      it("template", () => {
        const received = TestRenderer.create(
          <Boxl
            css={`
              color: red;
            `}
          />
        );
        expect(received).toMatchSnapshot();
      });

      it("object", () => {
        const received = TestRenderer.create(<Boxl css={{ color: "red" }} />);
        expect(received).toMatchSnapshot();
      });

      it("style function", () => {
        const received = TestRenderer.create(
          <Boxl css={s => s`color: red;`} />
        );
        expect(received).toMatchSnapshot();
      });

      it("style function w/ props interpolated", () => {
        const B = b.div<{ color: string }>();

        const received = TestRenderer.create(
          <B color="red" css={s => s`color: ${props => props.color}`} />
        );
        expect(received).toMatchSnapshot();
      });
    });
  });
});
