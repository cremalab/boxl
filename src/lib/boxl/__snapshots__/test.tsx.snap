// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`boxl childWrap auto 1`] = `
<div
  className="sc-jAaTju fiApCr"
  data-name="Boxl"
>
  <div
    className="sc-jDwBTQ hkTPWS"
    data-name="BoxlChild"
  >
    <div>
      1
    </div>
  </div>
  <div
    className="sc-jDwBTQ hkTPWS"
    data-name="BoxlChild"
  >
    <div>
      2
    </div>
  </div>
</div>
`;

exports[`boxl childWrap even 1`] = `
<div
  className="sc-iRbamj knnsep"
  data-name="Boxl"
>
  <div
    className="sc-jlyJG eDCHix"
    data-name="BoxlChild"
  >
    <div>
      1
    </div>
  </div>
  <div
    className="sc-jlyJG eDCHix"
    data-name="BoxlChild"
  >
    <div>
      2
    </div>
  </div>
  <div
    className="sc-jlyJG kwSobm"
    data-name="BoxlChild"
  />
  <div
    className="sc-jlyJG kwSobm"
    data-name="BoxlChild"
  />
</div>
`;

exports[`boxl direction default 1`] = `
<div
  className="sc-kEYyzF gdxGEn"
  data-name="Boxl"
>
  <div>
    1
  </div>
  <div>
    2
  </div>
</div>
`;

exports[`boxl direction horizontal 1`] = `
<div
  className="sc-jWBwVP kzeDgk"
  data-name="Boxl"
>
  <div>
    1
  </div>
  <div>
    2
  </div>
</div>
`;

exports[`boxl direction vertical 1`] = `
<div
  className="sc-hSdWYo hmUyiD"
  data-name="Boxl"
>
  <div>
    1
  </div>
  <div>
    2
  </div>
</div>
`;

exports[`boxl edge-case handles null child in children array 1`] = `
<div
  className="sc-chPdSV LgyEa"
  data-name="Boxl"
/>
`;

exports[`boxl edge-case naked text node is not wrapped when spacing undefined 1`] = `
<div
  className="sc-kpOJdX fROgiC"
  data-name="Boxl"
>
  <div>
    1
  </div>
  2
  <div>
    3
  </div>
</div>
`;

exports[`boxl edge-case naked text node is wrapped when spacing defined 1`] = `
<div
  className="sc-ckVGcZ ckgXKJ"
  data-name="Boxl"
>
  <div
    className="sc-jKJlTe gylQvj"
    data-name="BoxlChildren"
  >
    <div
      className="sc-eNQAEJ iAIntP"
      data-name="BoxlChild"
    >
      <div>
        1
      </div>
    </div>
    <div
      className="sc-eNQAEJ iAIntP"
      data-name="BoxlChild"
    >
      2
    </div>
    <div
      className="sc-eNQAEJ iAIntP"
      data-name="BoxlChild"
    >
      <div>
        3
      </div>
    </div>
  </div>
</div>
`;

exports[`boxl structure renders correct structure when no spacing 1`] = `
<div
  className="sc-bwzfXH RkOMy"
  data-name="Boxl"
>
  Text
</div>
`;

exports[`boxl structure renders correct structure when spacing AND children length > 1 1`] = `
<div
  className="sc-bZQynM hmYzAv"
  data-name="Boxl"
>
  <div
    className="sc-gzVnrw gviVII"
    data-name="BoxlChildren"
  >
    <div
      className="sc-htoDjs dCMzwX"
      data-name="BoxlChild"
    >
      <div>
        Text 1
      </div>
    </div>
    <div
      className="sc-htoDjs dCMzwX"
      data-name="BoxlChild"
    >
      <div>
        Text 2
      </div>
    </div>
  </div>
</div>
`;

exports[`boxl structure renders correct structure when spacing AND children length > 1 and children are Boxl 1`] = `
<div
  className="sc-dnqmqq ilEALG"
  data-name="Boxl"
>
  <div
    className="sc-iwsKbI fdOaWl"
    data-name="BoxlChildren"
  >
    <div
      className="sc-gZMcBi bLlThd"
      data-name="BoxlChild"
    >
      <div
        className="sc-VigVT faYjXQ"
        data-name="Boxl"
      >
        Text 1
      </div>
    </div>
    <div
      className="sc-gZMcBi bLlThd"
      data-name="BoxlChild"
    >
      <div
        className="sc-jzJRlG hXlUWV"
        data-name="Boxl"
      >
        Text 2
      </div>
    </div>
  </div>
</div>
`;

exports[`boxl structure renders correct structure when spacing BUT children length === 1 1`] = `
<div
  className="sc-ifAKCX gZjnqQ"
  data-name="Boxl"
>
  Text
</div>
`;

exports[`rerender maintains element instance between renders 1`] = `
<input
  class="sc-hzDkRC mzrWx"
  data-name="Boxl"
  placeholder="test1"
  value="My password is PASSWORD"
/>
`;

exports[`rerender maintains element instance between renders 2`] = `
<input
  class="sc-fMiknA kqBRb"
  data-name="Boxl"
  placeholder="test2"
  value="My password is PASSWORD"
/>
`;
